src/main/java/com/flightstats/hub/model/ChannelContentKey.java

```diff
@@ -25,10 +35,49 @@ public String toUrl() {
         return "channel/" + channel + "/" + contentKey.toUrl();
     }
 
-    public static ChannelContentKey fromUrl(String url) {
-        String after = StringUtils.substringAfter(url, "channel/");
-        String[] split = StringUtils.split(after, "/", 2);
-        return new ChannelContentKey(split[0], ContentKey.fromUrl(split[1]).get());
+    /**
+     * @param url Expects the format of ".+/channel/channelName/yyyy/mm/dd/hh/mm/ss/sss/hash"
+     */
+    public static ChannelContentKey fromResourcePath(String url) {
+        String channelPath = StringUtils.substringAfter(url, "channel/");
+        return fromChannelPath(channelPath);
+    }
+
+    /**
+     * @param path Expects the format of ".+/channelName/yyyy/mm/dd/hh/mm/[ss][sss][hash]"
+     */
+    public static ChannelContentKey fromSpokePath(String path) {
+        try {
+            String[] split = path.split("/");
+            String channel = split[split.length - 7];
+            String year = split[split.length - 6];
+            String month = split[split.length - 5];
+            String day = split[split.length - 4];
+            String hour = split[split.length - 3];
+            String minute = split[split.length - 2];
+            String second = StringUtils.substring(split[split.length - 1], 0, 2);
+            String millisecond = StringUtils.substring(split[split.length - 1], 2, 5);
+            String hash = StringUtils.substring(split[split.length - 1], 5);
+            String channelPath = Stream.of(channel, year, month, day, hour, minute, second, millisecond, hash).map(String::valueOf).collect(Collectors.joining("/"));
+            return fromChannelPath(channelPath);
+        } catch (IllegalArgumentException | ArrayIndexOutOfBoundsException e) {
+            logger.error("cannot build key from spoke path: " + path);
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    /**
+     * @param path Expects the format of "channelName/yyyy/mm/dd/hh/mm/ss/sss/hash"
+     */
+    public static ChannelContentKey fromChannelPath(String path) {
+        String[] split = StringUtils.split(path, "/", 2);
+        String channelName = split[0];
+        Optional<ContentKey> key = ContentKey.fromUrl(split[1]);
+        if (key.isPresent()) {
+            return new ChannelContentKey(channelName, key.get());
+        } else {
+            throw new IllegalArgumentException("cannot build key from path: " + path);
```

Comment:
Is this expected, or just really defensive?

Reply:
I tried to address this but it stems from Optional usage upstream that would blow up the scope to fix. I'm going to punt on solving the Optional creep. The current code at least stops it at this level and either returns values or throws exceptions.